src/url-loader/fs-url-resolver.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as pathlib from 'path';
+import {posix as posix} from 'path';
+import {format as urlLibFormat, Url} from 'url';
+import Uri from 'vscode-uri';
+
+import {parseUrl} from '../core/utils';
+import {FileRelativeUrl, PackageRelativeUrl, ScannedImport} from '../index';
+import {ResolvedUrl} from '../model/url';
+
+import {UrlResolver} from './url-resolver';
+
+const isWindows = process.platform === 'win32';
+
+/**
+ * Resolves package-relative urls to a configured root directory.
+ */
+export class FsUrlResolver extends UrlResolver {
+  protected readonly packageDir: string;
+  protected readonly packageUrl: ResolvedUrl;
+  constructor(
+      packageDir: string|undefined, private readonly hostname?: string) {
+    super();
+    this.packageDir =
+        normalizeFsPath(pathlib.resolve(packageDir || process.cwd()));
+    this.packageUrl =
+        this.brandAsResolved(Uri.file(this.packageDir).toString());
+    if (!this.packageUrl.endsWith('/')) {
+      this.packageUrl = this.brandAsResolved(this.packageUrl + '/');
+    }
+  }
+
+  resolve(
+      unresolvedHref: FileRelativeUrl|PackageRelativeUrl,
+      baseUrl: ResolvedUrl = this.packageUrl,
+      _import?: ScannedImport): ResolvedUrl|undefined {
+    const resolvedHref = this.simpleUrlResolve(unresolvedHref, baseUrl);
+    if (resolvedHref === undefined) {
+      return undefined;
+    }
+    const url = parseUrl(resolvedHref);
+    if (this.shouldHandleAsFileUrl(url)) {
+      return this.handleFileUrl(url, unresolvedHref);
+    }
+    return this.brandAsResolved(resolvedHref);
+  }
+
+  protected shouldHandleAsFileUrl(url: Url) {
+    if (this.hostname !== undefined && url.hostname === this.hostname) {
```

Comment:
don't we want to use the whole `host`? For the purposes of URLs as identifiers, same `hostname` but different `port` might as well be a different domain.